src/invocation/InvocationService.ts

```diff
@@ -322,43 +351,221 @@ export class InvocationService {
-        var deferred = Promise.defer<string>();
-        var invocation = new Invocation(this.client, request);
-        invocation.handler = handler;
+    start() {
+        this.client.getConnectionManager().on('connectionOpened', this.onConnectionAdded.bind(this));
+        this.client.getConnectionManager().on('connectionClosed', this.onConnectionRemoved.bind(this));
+        this.connectionRefreshTask = this.connectionRefreshHandler();
+    }
+
+    protected connectionRefreshHandler(): void {
+        if (this.isShutdown) {
+            return;
+        }
+        this.trySyncConnectToAllConnections().finally(() => {
+            this.connectionRefreshTask =
+                setTimeout(this.connectionRefreshHandler.bind(this), this.connectionRefreshTaskInterval);
+        });
+    }
+
+    onConnectionAdded(connection: ClientConnection) {
+        this.reregisterListenersOnConnection(connection);
+
+    }
+
+    onConnectionRemoved(connection: ClientConnection) {
+        this.removeRegistrationsOnConnection(connection);
+    }
+
+    onHeartbeatRestored(connection: ClientConnection): void {
+        var failedRegistrationsOnConn: Set<string> = this.failedRegistrations.get(connection);
+        failedRegistrationsOnConn.forEach((userKey: string, ignored: string) => {
+            this.invokeRegistrationFromRecord(userKey, connection);
+        });
+    }
+
+    reregisterListeners() {
+        var connections = this.client.getConnectionManager().getActiveConnections();
+        for (var connAddress in connections) {
+            this.reregisterListenersOnConnection(connections[connAddress]);
+        }
+    }
+
+    reregisterListenersOnConnection(connection: ClientConnection) {
+        this.activeRegistrations.forEach((registrationMap: Map<ClientConnection, ClientEventRegistration>, userKey: string) => {
+            if (registrationMap.has(connection)) {
+                return;
+            }
+            this.invokeRegistrationFromRecord(userKey, connection).then((eventRegistration: ClientEventRegistration) => {
+                registrationMap.set(connection, eventRegistration);
+            }).catch((e) => {
+                this.logger.warn('ListenerService', e);
+            });
+        }, this);
+    }
+
+    removeRegistrationsOnConnection(connection: ClientConnection) {
+        this.failedRegistrations.delete(connection);
+        this.activeRegistrations.forEach((registrationsOnUserKey: Map<ClientConnection, ClientEventRegistration>,
+                                          userKey: string) => {
+            var eventRegistration: ClientEventRegistration = registrationsOnUserKey.get(connection);
+            if (eventRegistration !== undefined) {
+                this.client.getInvocationService().removeEventHandler(eventRegistration.correlationId.toNumber());
+            }
+        });
+    }
+
+    invokeRegistrationFromRecord(userRegistrationKey: string, connection: ClientConnection): Promise<ClientEventRegistration> {
```

Comment:
is this method share any code with `registerListenerInternal` ?  A shared invoke method will increase readability, if possible.

Reply:
This method takes the record of a previously registered listener and invokes the registration again when a connection is restored whereas `registerListenerInternal` registers a register for the first time. `registerListenerInternal` creates records of listeners while this method uses those records. Invocation part is only `this.client.getInvocationService().invoke(invocation)` part. The actions taken at the result of invocation are also different. It is not worth extracting another method and sharing it.